src/realm/object-store/sync/subscribable.hpp

```diff
@@ -0,0 +1,141 @@
+        , m_token(std::move(other.m_token))
+        {
+            other.m_subscribable = nullptr;
+        }
+        Token& operator=(Token&& other) {
+            m_subscribable = std::move(other.m_subscribable);
+            m_token = std::move(other.m_token);
+            other.m_subscribable = nullptr;
+        }
+        Token(const Token&) = delete;
+        Token& operator=(const Token&) = delete;
+
+        ~Token()
+        {
+            if (m_subscribable) {
+                m_subscribable->unsubscribe(*this);
+            }
+        }
+
+        uint64_t value() const
+        {
+            return m_token;
+        }
+
+    private:
+        Subscribable* m_subscribable;
+        uint64_t m_token;
+
+        template <class U> friend struct Subscribable;
+    };
+
+    using Observer = std::function<void(const T&)>;
+
+    Subscribable() = default;
+    Subscribable(const Subscribable& other)
+    {
+        std::lock_guard<std::mutex> lock(other.m_mutex);
+        m_subscribers = other.m_subscribers;
+    }
+    Subscribable(Subscribable&& other)
+    {
+        std::lock_guard<std::mutex> lock(other.m_mutex);
+        m_subscribers = std::move(other.m_subscribers);
+    }
+    Subscribable& operator=(const Subscribable& other) {
+        std::lock_guard<std::mutex> lock(other.m_mutex);
+        m_subscribers = other.m_subscribers;
+    }
+    Subscribable& operator=(Subscribable&& other) {
+        std::lock_guard<std::mutex> lock(other.m_mutex);
+        m_subscribers = std::move(other.m_subscribers);
+    }
+    /// Subscribe to notifications for class type T. Any mutation to the T class
+    /// will trigger the observer. Notifying subscribers must be done manually
+    /// by the Subscribable.
+    /// @param observer callback to be called on mutation
+    /// @returns a token identifying the observer
+    [[nodiscard]] Token subscribe(Observer&& observer)
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+        static uint64_t m_token = 0;
+        m_subscribers.insert({m_token, std::move(observer)});
+        return Token {this, m_token++};
+    }
+
+    /// Unsubscribe to notifications for this Subscribable using the
+    /// token returned when calling `subscribe`.
+    /// @param token the token identifying the observer.
+    void unsubscribe(const Token& token)
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+        m_subscribers.erase(token.m_token);
+    }
+
+    /// A count of subscribers subscribed to class T.
+    /// @return the amount of subscribers subscribed to class T.
+    size_t subscribers_count() const
+    {
+        return m_subscribers.size();
+    }
+
+protected:
+    /// Emit a change event to all subscribers.
+    void emit_change_to_subscribers(const T& subject) const
+    {
+        std::lock_guard<std::mutex> lock(m_mutex);
+        for (const auto& [_, subscriber] : m_subscribers) {
+            subscriber(subject);
```

Comment:
This'll deadlock if the subscriber tries to unsubscribe. The simple fix for that is to make a copy of subscribers and drop the lock before iterating over the copy. The more complicated fix that makes unsubscribing work fully correctly is to do the convoluted things that CollectionNotifier does.